### ISSUE TITLE
How can I exclude a list of directories via BACKUP_EXCLUDE_REGEXP?

### ISSUE THREAD

**Issue by bverkron** [offen/docker-volume-backup#164]
**What are you trying to do?**
Unable to get a relatively simple regex pattern to work with BACKUP_EXCLUDE_REGEXP and I'm wonder if I'm misunderstanding how the regex is interpreted in regards to paths, etc.

Given the directory structure below I'm attempting to exclude many of the directories with this regex statement:
```
Logs|Diagnostics|Crash\ Reports|Cache|Media| Metadata|Codecs|Drivers
```

Directory structure of /backup/
```
/backup/config/Library/Application Support/Plex Media Server # ls -lalh
total 64K    
drwxr-xr-x   13 1000     1000        4.0K Oct 22 23:24 .
drwxr-xr-x    3 1000     1000        4.0K Oct 22 22:24 ..
-rw-------    1 1000     1000          42 Oct 22 23:24 .LocalAdminToken
drwxr-xr-x    7 1000     1000        4.0K Oct 22 23:31 Cache
drwxr-xr-x    6 1000     1000        4.0K Oct 22 23:24 Codecs
drwxr-xr-x   14 1000     1000        4.0K Oct 22 23:20 Crash Reports
drwxr-xr-x    2 1000     1000        4.0K Oct 22 23:20 Diagnostics
drwxr-xr-x    2 1000     1000        4.0K Oct 22 22:24 Drivers
drwxr-xr-x    3 1000     1000        4.0K Oct 22 23:24 Logs
drwxr-xr-x    3 1000     1000        4.0K Oct 22 23:20 Media
drwxr-xr-x    4 1000     1000        4.0K Oct 22 23:20 Metadata
drwxr-xr-x    7 1000     1000        4.0K Oct 22 22:25 Plug-in Support
drwxr-xr-x    3 1000     1000        4.0K Oct 22 23:20 Plug-ins
-rw-------    1 1000     1000         988 Oct 22 23:24 Preferences.xml
drwxr-xr-x    2 1000     1000        4.0K Oct 22 22:24 Updates
-rw-r--r--    1 1000     1000           3 Oct 22 23:24 plexmediaserver.pid
```

This results in *everything* being excluded. The backup looks like this when expanded (note the last folder is expanded and there's nothing in it)
![image](https://user-images.githubusercontent.com/19174503/197367374-2c998c89-d76f-4dcc-b7b3-5716901ec148.png)

If I add a slash and greedy match after the folder name it works overall but includes empty copies of the directories I want to exclude. Which makes sense since I'm telling it to just exclude everything _within_ those directories. Not a huge deal but I'd prefer to exclude those directories completely, not leave them empty.

```
Logs\/.*|Diagnostics\/.*|Crash\ Reports\/.*|Cache\/.*|Media\/.*|Codecs\/.*|Drivers\/.*
```

Curious why the first one with simple "or" operators doesn't work as I expect. I even tried shortening it to just `Media|Metadata` and the entire backup was still empty. This works in another backup container where I'm just doing `logs|somethingelse`. Not sure why it isn't working here.

**Comment by m90** [comment]
A few things:

- The regular expression is used to match every file in the tree under `BACKUP_SOURCES`: https://github.com/offen/docker-volume-backup/blob/907deecdd06557bf613111a2afd0436570bec60c/cmd/backup/script.go#L420-L433
- Go flavored Regular Expressions can be a bit tricky. I don't have the time to dive into your specific setup, but maybe this helps : https://github.com/google/re2/wiki/Syntax
- Empty directories are included on purpose, everything else would be a bug: https://github.com/offen/docker-volume-backup/issues/43
- Instead of creating a complex rule for excluding directories, have you thought of selectively mounting just the directories you want to include in your backup?

**Comment by bverkron** [comment]
Managed to figure some things out. I think some of it has to do with the fact that there's string interpolation happening at several levels. Portainer / Compose, shell variables, etc.

This works all the way through.

```
      BACKUP_EXCLUDE_REGEXP: "debug|backups|\\.log$$"
```

An extra `\` to escape the primary one
<img width="500" alt="image" src="https://user-images.githubusercontent.com/19174503/197927566-f0490d42-1620-41a0-baef-5383003bf0e4.png">

An extra `$` to escape the trailing $ at the end of the string for the Compose / Portainer level
<img width="284" alt="image" src="https://user-images.githubusercontent.com/19174503/197927106-21d066f0-f878-40f3-9685-ad090223ff9f.png">

Double quotes helped with an earlier issues but I didn't capture the error.

I think in some other cases, perhaps due to the lack of quotes around the whole string, it would pass at the compose / Portainer level but break elsewhere and cause it to not match anything... maybe. Only did enough experimenting to get something working, didn't go far enough to concretely determine root cause.

----

> Instead of creating a complex rule for excluding directories, have you thought of selectively mounting just the directories you want to include in your backup?

I did consider that for my other case but I'd rather err on the side of caution and exclude things I know for certain to be unnecessary and get _everything_ else than be too restrictive and miss something important. Even if I got it correct now there could be changes down the road that would introduce new or differently named files that are important and I may not know about and thus wouldn't adjust the mappings to capture.
